chia/consensus/blockchain.py

```diff
@@ -1069,61 +1070,62 @@ def seen_compact_proofs(self, vdf_info: VDFInfo, height: uint32) -> bool:
+        #              | :
+        # peak -> *    | : reorg_chain
+        #          \   / :
+        #           \ /  :
+        #            *  <- fork point
+        #         :  |
+        #  main   :  |
+        #  chain  :  |
+        #         :  |
+        #         :  * <- genesis
+
+        # If the block is not part of the main chain, we're on a fork, and we
+        # need to find the fork point
+        peak_block = self.try_block_record(header_hash)
+        if peak_block is None:
+            peak_block = await self.get_block_record_from_db(header_hash)
+        assert peak_block is not None
+        if self.height_to_hash(peak_block.height) != header_hash:
             peak: Optional[BlockRecord] = self.get_peak()
-            prev_block_record = await self.get_block_record_from_db(curr.prev_header_hash)
-            reorg_chain: Dict[uint32, bytes32] = {}
-            if prev_block_record is not None and peak is not None:
-                # Then we look up blocks up to fork point one at a time, backtracking
-                height_to_hash, _ = await lookup_fork_chain(
-                    self,
-                    (peak.height, peak.header_hash),
-                    (prev_block_record.height, prev_block_record.header_hash),
-                    self.constants,
-                )
-                reorg_chain.update(height_to_hash)
+            assert peak is not None
+            reorg_chain: Dict[uint32, bytes32]
+            # Then we look up blocks up to fork point one at a time, backtracking
+            reorg_chain, _ = await lookup_fork_chain(
+                self,
+                (peak.height, peak.header_hash),
+                (peak_block.height, peak_block.header_hash),
+                self.constants,
+            )
 
-            for ref_height in block.transactions_generator_ref_list:
-                if ref_height in additional_height_dict:
-                    ref_block = additional_height_dict[ref_height]
-                    if ref_block.transactions_generator is None:
-                        raise ValueError(Err.GENERATOR_REF_HAS_NO_GENERATOR)
-                    result.append(bytes(ref_block.transactions_generator))
-                elif ref_height in reorg_chain:
+            to_remove = []
+            for ref_height in generator_refs:
+                if ref_height in reorg_chain:
                     gen = await self.block_store.get_generator(reorg_chain[ref_height])
                     if gen is None:
                         raise ValueError(Err.GENERATOR_REF_HAS_NO_GENERATOR)
-                    result.append(gen)
-                else:
-                    [gen] = await self.block_store.get_generators_at([ref_height])
-                    result.append(gen)
-        assert len(result) == len(ref_list)
-        return BlockGenerator(block.transactions_generator, result)
+                    generators[ref_height] = gen
+                    to_remove.append(ref_height)
```

Comment:
looks like you can remove from generator_refs inside the loop right after adding to generators i think it would be nicer

Reply:
I cannot remove items from `generator_refs` while iterating over it, see line 1114

Reply:
in that case you can change to_remove to to_validate and append the misses to it, that would be shorter / more readable but nit-picky not critical

Reply:
right, i missed that, nitpick but this is my suggestion

```suggestion
            remain_to_validate = []
 		    for ref_height in generator_refs:
             	if ref_height in reorg_chain:
                   gen = await self.block_store.get_generator(reorg_chain[ref_height])
                   if gen is None:
                       raise ValueError(Err.GENERATOR_REF_HAS_NO_GENERATOR)
                   generators[ref_height] = gen                   
                   continue
                
                remain_to_validat.append(ref_height)
            if len(remain_to_validate) > 0:
	            # any remaining references fall in the main chain, and can be looked up
	            # in a single query
	            generators.update(await self.block_store.get_generators_at(remain_to_validate))```

Reply:
it would be a bit more code than that. `remain_to_validate` would have to be populated in the `else` case as well. Alternatively, `generator_refs` is cloned at the top and then items are removed.